lib/src/package_config_json.dart

```diff
@@ -0,0 +1,243 @@
+    [Map<String, dynamic> /*?*/ extraData]) {
+  var bytes = file.readAsBytesSync();
+  for (int i = 0; i < bytes.length; i++) {
+    var char = bytes[i];
+    if (char == 0x20 || char == 0x09 || char == 0x0a || char == 0x0d) {
+      // Skip whitespace.
+      continue;
+    }
+    if (char == 0x7b /*{*/) {
+      // Probably JSON, definitely not .packages.
+      return parsePackageConfigBytes(bytes, file, extraData);
+    }
+    break;
+  }
+  // File is not JSON.
+  var alternateFile = File(
+      path.join(path.dirname(file.path), ".dart_tool", "package_config.json"));
+  if (alternateFile.existsSync()) {
+    return parsePackageConfigBytes(
+        alternateFile.readAsBytesSync(), alternateFile, extraData);
+  }
+  return _parseDotPackagesConfig(bytes, file);
+}
+
+PackageConfig readPackageConfigJsonFile(File file,
+        [Map<String, dynamic> /*?*/ extraData]) =>
+    parsePackageConfigBytes(file.readAsBytesSync(), file, extraData);
+
+PackageConfig readDotPackagesFile(File file) =>
+    packagesfile.parse(file.readAsBytesSync(), Uri.file(file.path));
+
+PackageConfig parsePackageConfigBytes(Uint8List bytes, File file,
+    [Map<String, dynamic> /*?*/ extraData]) {
+  // TODO(lrn): Make this simpler. Maybe parse directly from bytes.
+  return parsePackageConfigJson(
+      json.fuse(utf8).decode(bytes), Uri.file(file.path), extraData);
+}
+
+/// Creates a [PackageConfig] from a parsed JSON-like object structure.
+///
+/// The [json] argument must be a JSON object (`Map<String, dynamic>`)
+/// containing a `"configVersion"` entry with an integer value in the range
+/// 1 to [PackageConfig.maxVersion],
+/// and with a `"packages"` entry which is a JSON array (`List<dynamic>`)
+/// containing JSON objects which each has the following properties:
+///
+/// * `"name"`: The package name as a string.
+/// * `"rootUri"`: The root of the package as a URI stored as a string.
+/// * `"packageUri"`: Optionally the root of for `package:` URI resolution
+///     for the package, as a relative URI below the root URI
+///     stored as a string.
+/// * `"languageVersion"`: Optionally a language version string which is a
+///     an integer numeral, a decimal point (`.`) and another integer numeral,
+///     where the integer numeral cannot have a sign, and can only have a
+///     leading zero if the entire numeral is a single zero.
+///
+/// All other properties are ignored.
+///
+/// The [baseLocation] is used as base URI to resolve the "rootUri"
+/// URI referencestring.
+PackageConfig parsePackageConfigJson(dynamic json, Uri baseLocation,
+    [Map<String, dynamic> /*?*/ extraData]) {
+  if (!baseLocation.hasScheme || baseLocation.isScheme("package")) {
+    throw ArgumentError.value(baseLocation.toString(), "baseLocation",
+        "Must be an absolute non-package: URI");
+  }
+  if (!baseLocation.path.endsWith("/")) {
+    baseLocation = baseLocation.resolveUri(Uri(path: "."));
+  }
+  var map = json;
+  if (map is Map<String, dynamic>) {
```

Comment:
nit: remove most of the nesting here by adding a helper function instead, something like this:

```
// [checkValue] should throw if `value` isn't valid
T check<T>(dynamic value, void Function(dynamic) checkValue) {
  checkValue(value);
  return value as T;
}
```

This will also make the exceptions a lot easier to correlate with the checks (they will be right next to the conditions instead of far down in a deeply nested structure).

Reply:
Not sure I need the `checkValue` function. An entry is either missing or not the right type. Anything else happens later.